src/core/mac/sub_mac.cpp

```diff
@@ -612,6 +627,15 @@ void SubMac::SignalFrameCounterUsedOnTxDone(const TxFrame &aFrame)
 
     OT_UNUSED_VARIABLE(allowError);
 
+#if OPENTHREAD_CONFIG_WAKEUP_COORDINATOR_ENABLE
+    if (aFrame.GetType() == Frame::kTypeMultipurpose && aFrame.GetFrameCounter(frameCounter) == kErrorNone)
+    {
+        // As long as Multipurpose frames are encrypted in SubMac, sync the frame counter used by the radio layer after
+        // a Multipurpose frame transmission.
+        Get<Radio>().SetMacFrameCounter(frameCounter + 1);
```

Comment:
Again regarding managing the frame counter, here, we are calling `Get<Radio>().SetMacFrameCounter(frameCounter + 1)` to synchronize the radio frame counter, but the radio may have already used counters ahead of this one, and we can forcibly move it back.

Perhaps it would be better to use `SetMacFrameCounterIfLarger()` here, though I am not sure it will handle all race conditions.